File: Easy/643-maximum-average-subarray-i/try-8.py
/*
 * Problem #643: Maximum Average Subarray I
 * Difficulty: Easy
 * Submission: Try 8
 * status: Accepted
 * Language: python3
 * Date: 07/12/2024, 12:35:54
 * Link: https://leetcode.com/problems/maximum-average-subarray-i/
 */

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        window=0
        for i in range(len(nums)):
            if i<k:
                window+=nums[i]
                mx=window
            else:
                mx=max(mx,window:=(window+nums[i]-nums[i-k]))
        return mx/k

File: Medium/53-maximum-subarray/try-8.cpp
/*
 * Problem #53: Maximum Subarray
 * Difficulty: Medium
 * Submission: Try 8
 * status: Accepted
 * Language: cpp
 * Date: 06/12/2025, 16:04:23
 * Link: https://leetcode.com/problems/maximum-subarray/
 */

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans = INT_MIN;
        int bestValueL = INT_MAX;

        int preSum = 0;
        for(int r = 0;r < nums.size(); r++){
            bestValueL = min(bestValueL, preSum);
            ans = max(preSum + nums[r] - bestValueL, ans);
            preSum += nums[r];
        }
        return ans;
    }
};
